### PR TITLE
Implement `PartialEq` for all function pointers via a shim

### PR DESCRIPTION
**Pull request by oli-obk**
This allows comparing higher kinded function pointers, at least directly via `PartialEq`. `==` has too much magic to work with this.

We can implement this in libstd without needing a shim once https://github.com/rust-lang/rust/pull/99531 lands

this unblocks https://github.com/rust-lang/rust/pull/105750

r? types

### CODE REVIEW

File: tests/ui/fn/fn-ptr-eq.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+fn foo(_: &()) {}
+
+fn main() {
+    let x: for<'a> fn(&'a ()) = foo;
+    let y: for<'a> fn(&'a ()) = foo;
+    bar(x, y);
+
+    let x: fn(&'static ()) = foo;
+    let y: for<'a> fn(&'a ()) = foo;
+    x == y;
```
Comment (RalfJung) [inline tests/ui/fn/fn-ptr-eq.rs:12]: So it works if the RHS is higher-ranked, but not if the LHS is higher-ranked? That is odd...

File: compiler/rustc_mir_build/src/build/matches/test.rs [hunk 2]
```diff
@@ -837,15 +837,9 @@ fn trait_method<'tcx>(
     method_name: Symbol,
     substs: impl IntoIterator<Item = impl Into<GenericArg<'tcx>>>,
 ) -> ConstantKind<'tcx> {
```
Comment (lcnr) [inline compiler/rustc_mir_build/src/build/matches/test.rs:839]: why does this not use `lang_items`?

File: tests/ui/fn/fn-ptr-eq.rs (hunk 1 above)
Comment (oli-obk) [reply tests/ui/fn/fn-ptr-eq.rs:12]: that seems definitely fixable, but I think it's a preexisting issue of binary ops.

Comment (RalfJung) [reply tests/ui/fn/fn-ptr-eq.rs:12]: I assume this is related to this remark?

> == has too much magic to work with this.

I don't I understand what the magic is, I thought `a == b` was just `PartialEq::eq(&a, &b)`...

**Comment by oli-obk** [comment]
closing in favor of https://github.com/rust-lang/rust/pull/108080

**Pull request closed without merging**